// parse-server/lib/graphql/types/QueryConstraint.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QueryConstraint = undefined;

var _graphql = require('graphql');

const supportedOperators = ['$eq', '$ne', '$in', '$nin', '$exists', '$select', '$dontSelect'];

const QueryConstraint = exports.QueryConstraint = new _graphql.GraphQLScalarType({
  name: 'QueryConstraint',
  description: `Generic Query Constraint
  Supported constraints:
  
  Equal To:
  - key: "value"
  - key: {$eq: "value"}
  
  Not Equal To
  - key: {$ne: "value"}
  
  Contained in:
  - key: {$in: ["value1", "value2"]}
  Not Contained in:
  - key: {$nin: ["value1", "value2"]}
  
  Exists: 
  - key: {$exists: true}
  
  This matches a value for a key in the result of a different query
  - key: {$select: {"query": {"className":"Team","where":{"winPct":{"$gt":0.5}}},"key":"city"}}}

  Requires that a key’s value not match a value for a key in the result of a different query
  - key: {$dontSelect: {"query": {"className":"Team","where":{"winPct":{"$gt":0.5}}},"key":"city"}}}
  `,
  serialize: () => {
    throw "StringQuery serialize not implemented";
  },
  parseValue: () => {
    throw "StringQuery parseValue not implemented";
  },
  parseLiteral: ast => {
    if (ast.kind == _graphql.Kind.OBJECT) {
      const fields = ast.fields;
      return fields.reduce((memo, field) => {
        const operator = field.name.value;
        if (supportedOperators.indexOf('$' + operator) > -1) {
          const value = field.value.value;
          memo['$' + operator] = value;
        }
        return memo;
      }, {});
    } else if (ast.kind == _graphql.Kind.STRING) {
      return ast.value;
    } else {
      throw 'Invalid literal for StringQuery';
    }
  }
});
